### ISSUE TITLE
"Identity not found" error with managed identity in GitHub Action

### ISSUE THREAD

**Issue by ajaskiewiczpl** [hashicorp/terraform-provider-azurerm#27927]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.9.8

### AzureRM Provider Version

3.114.0

### Affected Resource(s)/Data Source(s)

provider "azurerm"

### Terraform Configuration Files

```hcl
GitHub Action Job:

      - name: Terraform
        shell: pwsh
        working-directory: ./terraform
        run: |
            terraform --version
            terraform init -backend-config="key=${{secrets.TERRAFORM_STORAGE_KEY}}" -input=false
        env:
          ARM_USE_MSI: true
          ARM_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
          ARM_SUBSCRIPTION_ID: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
          ARM_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
```

Terraform config:
```
terraform {
  backend "azurerm" {
    resource_group_name  = "my-resource-group"
    storage_account_name = "mystorage"
    container_name       = "tfstateapi"
  }

  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
    }
  }
}

provider "azurerm" {
  features {
    key_vault {
      purge_soft_delete_on_destroy    = true
      recover_soft_deleted_key_vaults = true
    }
  }
}
```
```


### Debug Output/Panic Output

```shell
Run terraform --version
  terraform --version
  terraform init -backend-config="key=***" -input=false
  shell: /usr/bin/pwsh -command ". '{0}'"
  env:
    TERRAFORM_CLI_PATH: /home/runner/work/_temp/d8221f80-c6cb-4769-a9d0-df14aeba9d7c
    ARM_USE_MSI: true
    ARM_TENANT_ID: ***
    ARM_SUBSCRIPTION_ID: ***
    ARM_CLIENT_ID: ***
  
Terraform v1.9.8
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v3.114.0
+ provider registry.terraform.io/hashicorp/random v3.6.2
Initializing the backend...
╷
│ Error: Failed to get existing workspaces: Error retrieving keys for Storage Account "mystorage": azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/***/resourceGroups/my-resource-group/providers/Microsoft.Storage/storageAccounts/mystorage/listKeys?api-version=2021-01-01: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"Identity not found"} Endpoint http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&client_id=***&resource=https%3A%2F%2Fmanagement.azure.com%2F
│ 
│ 
╵
Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```

### References

This was supposed to be fixed in v3.84.0 according to https://github.com/hashicorp/terraform-provider-azurerm/issues/24172

**Comment by ajaskiewiczpl** [comment]
Yes, I have configured managed identity. In that github action, right before the `Terraform` step, I have azure login step and this works perfectly fine:
```
      - name: Azure Login
        uses: azure/login@v2
        with:
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
```

After reading this comment https://github.com/hashicorp/terraform-provider-azurerm/issues/24172#issuecomment-1849255484, my understanding is that the terraform provider uses wrong URL. This is exactly the same case as mine. But anyway, I will try to use OIDC instead.

**Comment by ajaskiewiczpl** [comment]
OIDC wasn't straightforward as well. Documentation lacks information about required environment variable `ARM_USE_OIDC: true` I would never guess by myself as I followed the docs. I found solution in other issues: https://github.com/hashicorp/terraform-provider-azurerm/issues/27490 and https://github.com/hashicorp/terraform-provider-azurerm/issues/22034#issuecomment-2239663241

**Comment by magodo** [comment]
Yes, the endpoint under used by the backend is deprecated. The auth related code in the azurerm backend (i.e. at the terraform core) is quite out dated. The work to upgrade to match the provider is tracked at https://github.com/hashicorp/terraform/issues/34322. There is an attempt from the team [here](https://github.com/hashicorp/terraform/pull/35381), but is closed.

**Comment by katbyte** [comment]
Thank you for asking this question!

It looks like this was answered by @magodo [in the thread](https://github.com/hashicorp/terraform-provider-azurerm/issues/27927#issuecomment-2463412666), so as part of a cleanup of historical issues we are going to close this out as answered.

For provider usage questions, the [HashiCorp Discuss forum](https://discuss.hashicorp.com/c/terraform-providers/31) tends to be the best place to get help. If this has turned out to be a bug or a missing feature on the current version of the provider (v5.x), please open a new issue with fresh details referencing this one. Thank you!
